### DISCUSSION TITLE
Frequency responses of various hifiberry cards

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#182] [General]
Hello all, I have finally got around to testing the frequency response of various hifiberry cards with my tweeters. The results are encouraging. 

As background, for a while we have thought there is a lowpass filter applied to the output of the hifiberry, to remove signals above some cutoff frequency. This makes sense for human listeners but not necessarily for mice.
https://wiki.auto-pi-lot.com/index.php/HiFiBerry_Amp2
https://twitter.com/chrisXrodgers/status/1535849926533562368
https://support.hifiberry.com/hc/en-us/community/posts/201846541-Ultrasonic-Frequency-Response-
https://support.hifiberry.com/hc/en-us/community/posts/360014295118-Ultrasonic-frequency-response

I tested the output of 3 cards:
Hifiberry Amp2 https://www.hifiberry.com/shop/boards/hifiberry-amp2/
Hifiberry DAC2 Pro https://www.amazon.com/gp/product/B08GKGVMZN
Hifiberry DAC2 HD https://www.amazon.com/gp/product/B08C1ZMDF1
I'm using Amazon links to the actual products I bought because Hifiberry's website sometimes seems to use DAC2 and DAC+ interchangeably, and I am confused about the difference. In particular the DAC2 Pro I have doesn't really match any of the DACs on the Hifiberry website, but is most similar visually to the DAC+ Pro.

I played full bandwidth white noise using Autopilot. I measured the audio power using an HBK ultrasonic microphone that is guaranteed flat up to 100 kHz. I also measured the voltage on the output of the DAC.

I tested two speakers, neither of which are available anymore. They were dirt cheap and promoted as ultrasonic transducers.
Almencla 2-inch 2.5-60KHz Piezo Horn Speaker Tweeter
Multicomp MCPCT-G5100-4139
I also made measurements with no speaker connected. Obviously no sound was played, but I could still measure the voltage coming out of the DAC. Importantly (spoiler) this voltage was very different with a speaker connected versus not.

First let's just look at the voltage output of each DAC. See figure here: https://www.dropbox.com/s/08lqlswf70jl3b0/dac_comparison.svg?dl=1
* Top subplot. The DAC2 Pro (labeled dacplus) has the flattest frequency response. The DAC2 HD (labeled dacplushd) is okay for low frequencies but starts falling off above 30 kHz. The Amp2 (blue) actually shows a resonant peak at 65kHz, and does not start falling off until 90 kHz. 
* Middle subplot. When we attach a speaker, things change. Now the DAC2 Pro and DAC2 HD both strongly attenuate at high frequencies. They are not capable of driving a speaker well at high frequencies. The Amp2, in contrast, maintains a fairly similar response even with a speaker attached.
* Bottom subplot: quantifies attenuation to voltage output when speaker is attached.

Next let's take a look at the audio power coming out of the speaker in each condition. See figure here:
https://www.dropbox.com/s/u7a9huc51xa8mxj/audio_power_analysis.svg?dl=0
* Top and middle subplots: the audio power for each speaker. Results are similar for both speakers. The DAC2 Pro and DAC2 HD are unable to drive this speaker well above 25 kHz or so. The Amp2 does a great job up until 75 kHz or so. 

So in conclusion, at high frequencies, it becomes more difficult to apply a voltage to the speakers, and the audio they produce for a given voltage is lower. But the Amp2 effectively compensates for this, first because it is better able to drive the speakers at high frequencies, and second because its own voltage output actually increases at high frequencies. 

I see no evidence for a lowpass filter on the output of the Amp2, though I do definitely see one on the DAC2 HD. This matches what Hifiberry support wrote in this comment on Dec 2020 (https://support.hifiberry.com/hc/en-us/community/posts/360014295118/comments/360003504057). Is there any documented evidence of a lowpass filter at 20 kHz on the Amp2? 

However these results are very dependent on the load: you will get a very different result if you test just with no speaker (open circuit), a load resistor (purely resistive), or different kinds of speakers. I certainly could believe that there are tweeters out there that the Amp2 cannot drive but other, specialized amplifiers can drive (perhaps Ralph's are like this). They probably just become very low-impedance at high frequencies and the Amp2 doesn't have enough current to drive them. My cheap tweeters don't seem to suffer from this problem until 75 kHz or so.

Next I am going to apply some filtering to the audio stimuli in order to produce sound out of the speaker that is flat from 5 kHz to 75 kHz, I think this should be possible.

**Comment by sneakers-the-rat** [comment]
Chris this rocks!!!!!! Adding a wiki tag to remember to add it there. Good to know that the frequency response changes depending on whether the resistor is a speaker vs. a resistor (i shoulda known that the speaker wouldn't be *totally* passive)

The only evidence we have for the output filter on the amp is from the support forum as well as there seeming to literally be an LC filter visible on the chip, though I didn't measure the capacitance/inductance of either of the components when I took them off.

**Reply by cxrodgers** [reply]
Thanks, and thanks for adding to wiki!

When you mentioned LC filter, it reminded me that these filters have a resonant frequency, which could explain the 65 kHz peak in my measurements. In fact the response I measured looks similar to some of the plots in this article (https://www.eetimes.com/understanding-output-filters-for-class-d-amplifiers/) especially Figs 4 and 5. If so, the lowpass cutoff of the LC filter they used is over 70 kHz.

The support forum hits I found seem to be discussing the DAC2 HD, which indeed has a much lower cutoff frequency, around 25 kHz.

**Reply by sneakers-the-rat** [reply]
I've never designed an LC filter, but it seems like it would be as straightforward as an RC filter????? (ducks tomatoes thrown by engineers who pick up the scent of me being wrong about electronics as usual)

I had made a quick note on the wiki page for the Amp2 ( https://wiki.auto-pi-lot.com/index.php/HiFiBerry_Amp2 ) and yeah it looks like that was just for the [DAC2](https://support.hifiberry.com/hc/en-us/community/posts/360014295118-Ultrasonic-frequency-response?mobile_site=true). 

Here's what I tried with the Amp2, in which I got results similar to yours. The big black boxes are inductors and the tiny lil things are capacitors, so I just pried them off one channel, which gave me really massive aliasing artifacts. At least they sure look like inductors, but the board says "TR1" for the part ID (the capacitors aare C21, C22, etc.), so not sure if that indicates it's something else

![shorting_LC_amp2-01](https://user-images.githubusercontent.com/12961499/182003465-e2e3508d-22d9-48d5-b034-489f9a7552a4.png)

Which has the 2nd-order LC filter suggested by the datasheet (linked on wiki) for the amplifier IC (although now that i'm looking at it more closely there's another capacitor I missed on the IC side of the inductor.

<img width="966" alt="Screen Shot 2022-07-30 at 3 52 06 PM" src="https://user-images.githubusercontent.com/12961499/182003495-5b552a78-eb84-4504-b601-7473f9422ff2.png">

I don't have my stuff setup to measure the inductance and can't find a match for the numbers on it (100 M582), but the capacitor is ~5.6uF (measured while still on the board tho) 

So if the [LC equation](https://electronicbase.net/low-pass-filter-calculator/) is 

<img width="343" alt="Screen Shot 2022-07-30 at 4 43 35 PM" src="https://user-images.githubusercontent.com/12961499/182003694-9238c0be-a035-4efd-8fe3-1fadec8c4660.png">

and the cutoff is 65kHz then that would make the inductor 1.07uH.

Not sure how to optimally design the filter if we'd want to raise it, but for example [this tool](https://rf-tools.com/lc-filter/) says 186pF and 5nH
<img width="1209" alt="Screen Shot 2022-07-30 at 4 45 55 PM" src="https://user-images.githubusercontent.com/12961499/182003771-0fdaee49-dd5f-4bac-b485-3faae9c584af.png">

**Comment by jmdelahanty** [comment]
I think something that will tickle you both happened recently in the lab I'm in. We threw an oscilloscope onto our speakers' wires to see why were getting these odd crackling noises from the `tone` function the Arduino outputs. We saw several things related to how the `tone` waveforms are generated, with the most important one being that the tones that are generated are very much square waves it seems like and also that there's a fair bit of jitter in the tone frequency too! An electrical engineer/HPC designer in our lab told us that we would need to either find a different driver for producing tones that's not from the Arduino (I did mention Autopilot again...) and/or adding an R/C component to the circuit. He also told us that we would very likely be producing harmonics of the tone we actually want!

Jonny told me about the perils of this a while ago. I figured it would be interesting to share an occurrence of looking at this stuff "from the field". And we're only producing tones that are 2 and 9 kHz...

**Reply by cxrodgers** [reply]
If I remember correctly, I think the Arduino generates tones with PWM. That's because it has only digital outputs, not analog outputs -- so the best it can do is generate a square wave. The jitter you mentioned may actually be the PWM at work. If you lowpass filter it, you should get a reasonable-looking sine wave. But yes, for better audio output, you need something like a sound card or an analog output device.

**Reply by jmdelahanty** [reply]
That is what it does (PWM bit)! Is that kind of filtering literally what an audio card does? I've never really looked into what they're up to.

**Comment by cxrodgers** [comment]
That is how a Class-D audio amplifier works (or, as far as I understand
it). There is a lot more to the details of it, which is why people don't
replace their stereos with Arduinos, but I think that's the general idea.
https://en.wikipedia.org/wiki/Class-D_amplifier
